sdcm/utils/url_utils.py

```diff
@@ -0,0 +1,83 @@
+import re
+from collections import namedtuple
+from urllib.request import urlopen
+
+from sdcm.utils.common import ParallelObject
+
+SCYLLA_URL_RESPONSE_TIMEOUT = 30
+SCYLLA_URL_REGEX = re.compile('.*(?P<url>http.*scylladb.com.*).*')
+SUPPORTED_FILE_EXTENSIONS = ('list', 'repo', 'xml', 'Packages', 'gz')
+
+BASE_ARCH_KEY = '$basearch'
+RELEASE_VERSION_KEY = '$releasever'
+URL_ENV_VARIABLES = {
+    BASE_ARCH_KEY: ['x86_64', 'noarch'],
+    RELEASE_VERSION_KEY: ['7'],
+}
+
+UrlDetails = namedtuple("UrlDetails", "is_xml_file urls")
+UrlDetails.__new__.__defaults__ = (None,) * len(UrlDetails._fields)
+
+
+def validate_url_is_file_and_extract_scylla_urls(urls, validate_file_output=True):
+    def _validate_url(_url):
+        is_xml_file = False
+        if not SCYLLA_URL_REGEX.match(_url):
+            return []
+
+        file_extension = _url.rsplit('/', maxsplit=1)[1]
+        if file_extension.endswith(('xml', 'xml.gz',)):
+            is_xml_file = True
+        if not file_extension.endswith(SUPPORTED_FILE_EXTENSIONS):
+            raise TypeError(f"The file extension '{file_extension}' not supported")
+        response = urlopen(url=_url)
+        response_status = response.status
+        if response_status != 200:
+            raise ValueError(f"The URL '{_url}' is invalid (got status code '{response_status}')")
+        if not validate_file_output:
+            return []
+
+        response_data = response.read().decode("utf-8")
+        if not response_data:
+            raise ValueError(f"The file that downloaded from url '{_url}' is empty")
+        return is_xml_file, list({line for line in response_data.split('\n') if SCYLLA_URL_REGEX.match(line)})
+
+    if not isinstance(urls, set):
+        urls = set(urls) if not isinstance(urls, str) else {urls}
+
+    for url in urls:
+        if not SCYLLA_URL_REGEX.match(url):
+            raise ValueError(f"The following URL not supported '{url}'")
+    result = [res.result for res in ParallelObject(objects=urls, timeout=SCYLLA_URL_RESPONSE_TIMEOUT).run(
```

Comment:
please take the parallel logic out of this function, and use it only where it's needed.

don't change `get_branch_version` cause of this, it make the code calling it not that clear, see the unittest that were changed cause of it.

Reply:
I've create function for parallel logic and the function `get_branch_version` returns only one value